tls/s2n_recv.c

```diff
@@ -127,13 +142,11 @@ ssize_t s2n_recv_impl(struct s2n_connection *conn, void *buf, ssize_t size, s2n_
         uint8_t record_type;
         int r = s2n_read_full_record(conn, &record_type, &isSSLv2);
         if (r < 0) {
-            if (s2n_errno == S2N_ERR_CLOSED) {
-                *blocked = S2N_NOT_BLOCKED;
-                if (!bytes_read) {
-                    return 0;
-                } else {
-                    return bytes_read;
-                }
+            /* Don't propagate the error if we already read some bytes.
+             * We'll report S2N_ERR_CLOSED on the next call.
+             */
+            if (s2n_errno == S2N_ERR_CLOSED && bytes_read) {
```

Comment:
Do you need to set the blocked status for this case?

Reply:
That's a good question.

I think returning S2N_BLOCKED_ON_READ (like we do if we don't change it here) is correct. If this code is triggered, that means that we attempted another read, which means that the caller requested more data than we're returning with bytes_read. And we do really want them to call again in that case, because we want to give them the S2N_ERR_CLOSED error.